### PR TITLE
[Connections] Consider null and unknown as valid in url string validator

### PR DESCRIPTION
**Pull request by fridgepoet**
An [issue](https://github.com/grafana/support-escalations/issues/13779) was raised that string validation was returning an error on the result of `each.value` from `for_each`. In the following example, existing string validation was rejecting `each.value` because it is null/unknown. 
This PR allows for null/unknown values in our string validation and adds tests using the `for_each`.

Example:
```hcl
locals {
  jobs = [
    { name = "test", url = "<some valid url>" }
  ]
}

resource "grafana_connections_metrics_endpoint_scrape_job" "valid_url" {
  for_each = { for j in local.jobs : j.name => j.url }
  stack_id = "1"
  name = each.key
  enabled = true
  authentication_method = "bearer"
  authentication_bearer_token = "<some valid token>"
  url = each.value
}
```

Running `terraform plan` before would result in: 
```
╷
│ Error: value must be valid URL with HTTPS
│
│   with grafana_connections_metrics_endpoint_scrape_job.valid_url,
│   on main.tf line 19, in resource "grafana_connections_metrics_endpoint_scrape_job" "valid_url":
│   19:   url = each.value
│
│ A valid URL is required.
│
│ Given Value: ""
│ 
```

Running `terraform plan` with the changes in this PR now results in:
```
Terraform will perform the following actions:

  # grafana_connections_metrics_endpoint_scrape_job.valid_url["test"] will be created
  + resource "grafana_connections_metrics_endpoint_scrape_job" "valid_url" {
      + authentication_bearer_token = (sensitive value)
      + authentication_method       = "bearer"
      + enabled                     = true
      + id                          = (known after apply)
      + name                        = "test"
      + scrape_interval_seconds     = 60
      + stack_id                    = "1"
      + url                         = "<some valid url>"
    }

Plan: 1 to add, 0 to change, 0 to destroy.

```
Applying is successful and one can see the job in the UI:
Screenshot 2024-12-06 at 11 44 27

Fixes https://github.com/grafana/support-escalations/issues/13779

### CODE REVIEW

File: internal/resources/connections/metrics_endpoint_scrape_job_test.go [hunk 1]
```diff
@@ -130,3 +142,39 @@ resource "grafana_connections_metrics_endpoint_scrape_job" "test" {
...
+  stack_id = "......"
+  name = each.key
+  enabled = false
+  authentication_method = "bearer"
+  authentication_bearer_token = "test"
+  url = each.value
+}
+`
+
+var resourceWithForEachInvalidURL = `
```
Comment (matthewnolf) [inline internal/resources/connections/metrics_endpoint_scrape_job_test.go:164]: So with the [early break](https://github.com/grafana/terraform-provider-grafana/pull/1942/files#diff-9074a61dbecc623cb6d2571cc8efdf409bd20e2f72af09153891b16b665fa0a3R35-R37) on `IsNull` or `IsUnknown`, we still do the validation at a later point?

**Comment by Duologic** [comment]
I think the tests are fixed after a rebase.

File: internal/resources/connections/metrics_endpoint_scrape_job_test.go (hunk 1 above)
Comment (fridgepoet) [reply internal/resources/connections/metrics_endpoint_scrape_job_test.go:164]: Yes, this example provides an empty string so it continues to https://github.com/grafana/terraform-provider-grafana/blob/340c6908a98ced498eee023261220a682003d471/internal/resources/connections/resources.go#L42

(unit test on empty string: https://github.com/grafana/terraform-provider-grafana/blob/e5e3eb356e1d2e4c99b3d862dee1889ffd5ca9af/internal/resources/connections/resources_test.go#L34-L33)

**Review by matthewnolf — APPROVED** [review]
LGTM

**Pull request merged**